### PR TITLE
Fixed a crash on closing mangosd.exe

### PR DESCRIPTION
**Pull request by Psimage**
Crash log:

<pre>
Revision: * * 1809 *
Date 8:11:2011. Time 17:11 
//=====================================================
*** Hardware ***
Processor: Intel(R) Celeron(R) M CPU        520  @ 1.60GHz
Number Of Processors: 1
Physical Memory: 2087024 KB (Available: 1133052 KB)
Commit Charge Limit: 2701204 KB

*** Operation System ***
Microsoft Windows XP Professional Service Pack 3 (Version 5.1, Build 2600)

//=====================================================
Exception code: C0000005 ACCESS_VIOLATION
Fault address:  0090F629 02:000BB629 D:\MaNGOS\zero_dev\Mangos_Server\bin\Win32_Debug\mangosd.exe

Registers:
EAX:E5E5E5E9
EBX:020EB150
ECX:E5E5E5E5
EDX:78CC5600
ESI:02F9FE70
EDI:02F9FD04
CS:EIP:001B:0090F629
SS:ESP:0023:02F9FCBC  EBP:02F9FD10
DS:0023  ES:0023  FS:003B  GS:0000
Flags:00010246

Call stack:
Address   Frame     Function      SourceFile
0090F629  00000000  
</pre>

With some debug magic i found that BattleGroundMng tries to erase some not existing data.
<dt>New problem</dt>

After shutdown command (or restart), mangosd.exe proccess freezes (he still works but eats all CPU time) and not exiting by himself (if i click "close" button it closes process). Debug shows that it start freezing after "exit(code)" function (after executing main).

### CODE REVIEW

File: src/game/BattleGround.cpp [hunk 1]
```diff
@@ -271,7 +271,9 @@ BattleGround::~BattleGround()
         DelObject(i);
 
     sBattleGroundMgr.RemoveBattleGround(GetInstanceID(), GetTypeID());
-    sBattleGroundMgr.DeleteClientVisibleInstanceId(GetTypeID(), GetBracketId(), GetClientInstanceID());
+	// if its not template bg
+	if(GetBracketId()!=MAX_BATTLEGROUND_BRACKETS)
```
Comment (Psimage) [inline src/game/BattleGround.cpp]: For default object of BattleGround, GetBracketId() returns <b>MAX_BATTLEGROUND_BRACKETS</b>

File: src/game/BattleGround.cpp [hunk 2]
```diff
@@ -271,7 +271,9 @@ BattleGround::~BattleGround()
...
+		sBattleGroundMgr.DeleteClientVisibleInstanceId(GetTypeID(), GetBracketId(), GetClientInstanceID());
```
Comment (Psimage) [inline src/game/BattleGround.cpp:278]:
``` C++
void DeleteClientVisibleInstanceId(BattleGroundTypeId bgTypeId, BattleGroundBracketId bracket_id, uint32 clientInstanceID)
{
    m_ClientBattleGroundIds[bgTypeId][bracket_id].erase(clientInstanceID);
}

ClientBattleGroundIdSet m_ClientBattleGroundIds[MAX_BATTLEGROUND_TYPE_ID][MAX_BATTLEGROUND_BRACKETS];
```

Here we have out of bound exception (<b>bracked_id</b> was <b>MAX_BATTLEGROUND_BRACKETS</b>).

**Comment by DomGries** [comment]
http://getmangos.com/community/topic/16971/one-server-crashes-on-windows-reasonworkaround/ . Either a coincidence or you copied the patch from me. Either way the proper patch is in my last post in the linked thread :)

**Comment by Psimage** [comment]
Thx. Your version is nicer :)
But why its still not in release? And don't you know something how to solve the second problem?

**Pull request merged**